Add extension documentation (#3977)
powsybl/powsybl-core: docs/grid_model/extensions.md

(active-power-control-extension)=
## Active power control

> This extension is used to configure the participation factor of the generator, typically in the case of a load flow computation with distributed slack enabled (with [balance type](../simulation/loadflow/configuration.md#balanceType) on generator). This extension is attached to a [generator](network_subnetwork.md#generator) or a [battery](network_subnetwork.md#battery).

Reviewer: ```suggestion
This extension is used to configure the participation factor of the generator, typically in the case of a load flow computation with distributed slack enabled (with [balance type](../simulation/loadflow/configuration.md#balancetype) on generator). This extension is attached to a [generator](network_subnetwork.md#generator) or a [battery](network_subnetwork.md#battery).
```

(busbar-section-position-extension)=
## Busbar section position

This extension is attached to a [bus bar section](network_subnetwork.md#busbar-section) and gives positions information about a busbar section.

Note that a busbar is a set of busbar sections. Hence, the sections of the same busbar should have the same busbar index. The busbar indices induce an order of busbars within the voltage level, which usually reflects the busbars physical relative positions. Similarly, the section indices induce an order of sections of the same busbar, which usually reflects their physical relative position.

| Attribute             | Type | Unit | Required | Default value | Description                                                          |
|-----------------------|------|------|----------|---------------|----------------------------------------------------------------------|
| busbarIndex           | int  | -    | yes      | -1            | The position of the bus bar section relatively to others             |

> | sectionIndex          | int  | -    | yes      | -1            | The position of the bus bar section within the corresponding busbar  |

Reviewer: In the extension, these attributes are required and have no default values (only `0` and positive values are accepted, so -1 is rejected).

> Example of code to get the position and section position data for a bus bar section :

Reviewer: ```suggestion
Example of code to set the position and section position data for a bus bar section :
```

| value      | Object                         | -    | no       | -             | The discrete measurement value                                                                                   |
| valid      | boolean                        | -    | no       | -             | The validity status (if true, the discrete measured value cannot be null)                                        |

(dynamic-model-info-extension)=
## Dynamic Model Info
This extension is attached to an Identifiable. It is used to store the dynamic model used during a dynamic simulation.

| Attribute           | Type              | Unit | Required | Default value | Description                               |
|---------------------|-------------------|------|----------|---------------|-------------------------------------------|

> | modelNamePerVariant | ArrayList<String> | -    | no       | -             | A list of models used per network variant |

Reviewer: To be consistent with the other extensions (see for instance `HvdcOperatorActivePowerRange`):
```suggestion
| modelName | String | -    | no       | -             | The dynamic model used during a dynamic simulation |

This attribute is multi-variant: it can vary from one variant to another.
```

Reviewer: The type must also be fixed: `ArrayList<String>`  -> `String`

Author: Yes done :)

(entsoe-area-extension)=
## ENTSO-E area

This extension is attached to a [substation](network_subnetwork.md#substation) and is used to store its ENTSO-E geographical code.

> It is used by the UCTE and ENTSO-E data exchange formats. The code identifies the area the substation belongs to (for example `FR`, `BE`, or `D1`, `D2`... for the German areas) and is associated with a country.

Reviewer: ```suggestion
The code identifies the area the substation belongs to (for example `FR`, `BE`, or `D1`, `D2`... for the German areas) and is associated with a country.

When importing a UCTE file, this extension is created for the substations.
```

(reference-priority-extension)=
## Reference Priority

> This extension is attached to a [generator](network_subnetwork.md#generator), or a BusBarSection or a Load and is used to define the angle reference bus of

Reviewer: ```suggestion
This extension is attached to a [generator](network_subnetwork.md#generator), or a [busbar section](network_subnetwork.html#busbar-section) or a [load](network_subnetwork.html#load) and is used to define the angle reference bus of
```

Reviewer: Sorry, I was not attentive enough!
The links do not work: it is `network_subnetwork.md` and not `network_subnetwork.html`. I messed up with the extension!

Reviewer: Thanks!

The `targetV` and `participate` attributes are multi-variants: they can vary from one variant to another.

Here is how to add a secondary voltage control extension to a network:
```java
network.newExtension(SecondaryVoltageControlAdder.class)
    .newControlZone()
        .withName("z1")
        .newPilotPoint()
            .withBusbarSectionsOrBusesIds(List.of("NLOAD"))
            .withTargetV(15d)
        .add()
        .newControlUnit()
            .withId("GEN")
            .withParticipate(false)
        .add()
        .newControlUnit()
            .withId("GEN2")
        .add()
    .add()
.add();
```

This extension is provided by the `com.powsybl:powsybl-iidm-extensions` module.

(slack-terminal-extension)=
## Slack terminal

This extension is attached to a [voltage level](network_subnetwork.md#voltage-level) and is used to define the slack bus
of a power flow calculation i.e. which bus will be used to balance the active and reactive power in load flow analysis.
Use this extension before a computation to force the slack bus selection. You should enable default load flow parameter

> [`readSlackBus`](../simulation/loadflow/configuration.md#readSlackBus). Use this extension after a computation to attach

Reviewer: ```suggestion
[`readSlackBus`](../simulation/loadflow/configuration.md#readslackbus). Use this extension after a computation to attach
```

to the network the slack bus that has been selected by the load flow engine (one by connected component). You should enable

> default load flow parameter [`writeSlackBus`](../simulation/loadflow/configuration.md#writeSlackBus).

Reviewer: ```suggestion
default load flow parameter [`writeSlackBus`](../simulation/loadflow/configuration.md#writeslackbus).
```

This extension is provided by the `com.powsybl:powsybl-iidm-api` module.

(standby-automaton-extension)=
## Standby automaton

> This extension is attached to a [static var compensator](network_subnetwork.md#static-var-compensator) and is used to define the automaton status and its fixed part susceptance, high and low voltage setpoints and thresholds.

Reviewer: ```suggestion
This extension is attached to a [static VAR compensator](network_subnetwork.md#static-var-compensator) and is used to define the automaton status and its fixed part susceptance, high and low voltage setpoints and thresholds.
```

When the static VAR compensator is in standby mode, it behaves as a fixed shunt of susceptance `b0`. The high and low voltage setpoints and thresholds are used by the automaton to adjust the voltage regulation when the monitored voltage leaves the `[lowVoltageThreshold, highVoltageThreshold]` range.

| Attribute             | Type    | Unit | Required | Default value | Description                                                                                       |
|-----------------------|---------|------|----------|---------------|---------------------------------------------------------------------------------------------------|
| standby               | boolean | -    | true     | false         | The automaton status (in service or not)                                                          |

> | b0                    | double  | S    | true     | NaN           | The fix part of the susceptance (between the min and max f the static var compensator susceptance |

Reviewer: ```suggestion
| b0                    | double  | S    | true     | NaN           | The fix part of the susceptance (between the min and max f the static VAR compensator susceptance |
```

| lowVoltageSetpoint    | double  | kV   | true     | NaN           | The voltage setpoint used when the low voltage threshold is reached                               |
| highVoltageSetpoint   | double  | kV   | true     | NaN           | The voltage setpoint used when the high voltage threshold is reached                              |
| lowVoltageThreshold   | double  | kV   | true     | NaN           | The voltage threshold under which the static var compensator controls at low voltage setpoint     |

> | highVoltageThreshold  | double  | kV   | true     | NaN           | The voltage threshold above which the static var compensator controls at high voltage setpoint    |

Reviewer: ```suggestion
| lowVoltageThreshold   | double  | kV   | true     | NaN           | The voltage threshold under which the static VAR compensator controls at low voltage setpoint     |
| highVoltageThreshold  | double  | kV   | true     | NaN           | The voltage threshold above which the static VAR compensator controls at high voltage setpoint    |
```

The `standby`, `lowVoltageSetpoint`, `highVoltageSetpoint`, `lowVoltageThreshold` and `highVoltageThreshold` attributes are multi-variants: they can vary from one variant to another.

> Example of code to get the standby automaton data for a static var compensator :

Reviewer: ```suggestion
Example of code to add standby automaton data for a static VAR compensator :
```
